lib/sandbox.jail.ml

```diff
@@ -0,0 +1,162 @@
+    Lwt_unix.read src buf 0 (Bytes.length buf) >>= function
+    | 0 -> Lwt.return_unit
+    | n -> Build_log.write dst (Bytes.sub_string buf 0 n) >>= aux
+  in
+  aux ()
+
+let run ~cancelled ?stdin:stdin ~log (t : t) config rootdir =
+  let cwd = rootdir in
+  Lwt_mutex.with_lock t.lock (fun () ->
+  Os.with_pipe_from_child @@ fun ~r:out_r ~w:out_w ->
+  let rootdir = rootdir / "rootfs" in
+  let workdir = rootdir / config.Config.cwd in
+  (* Make sure the work directory exists prior to starting the jail. *)
+  begin
+    match Os.check_dir workdir with
+    | `Present -> Lwt.return_unit
+    | `Missing -> Os.sudo [ "mkdir" ; "-p" ; workdir ]
+  end >>= fun () ->
+  let stdout = `FD_move_safely out_w in
+  let stderr = stdout in
+  let copy_log = copy_to_log ~src:out_r ~dst:log in
+  let proc =
+    let cmd =
+      let options = jail_options config rootdir in
+      "jail" :: "-c" :: t.jail_name :: options
+    in
+    let stdin = Option.map (fun x -> `FD_move_safely x) stdin in
+    let pp f = Os.pp_cmd f ("", cmd) in
+    (* This is similar to
+       Os.sudo_result ~cwd ?stdin ~stdout ~stderr ~pp cmd
+       but also unmounting the in-jail devfs if necessary, see below. *)
+    let cmd = if Os.running_as_root then cmd else "sudo" :: "--" :: cmd in
+    Logs.info (fun f -> f "Exec %a" Os.pp_cmd ("", cmd));
+    !Os.lwt_process_exec ~cwd ?stdin ~stdout ~stderr ~pp
+      ("", Array.of_list cmd) >>= function
+    | Ok 0 ->
+      (* If the command within the jail completes, the jail is automatically
+         removed, but without performing any of the stop and release actions,
+         thus we can not use "exec.stop" to unmount the in-jail devfs
+         filesystem. Do this here, ignoring the exit code of umount(8). *)
+      let cmd = [ "sudo" ; "/sbin/umount" ; rootdir / "dev" ] in
+      Os.exec ~is_success:(fun _ -> true) cmd >>= fun () ->
+      Lwt_result.ok Lwt.return_unit
+    | Ok n -> Lwt.return @@ Fmt.error_msg "%t failed with exit status %d" pp n
+    | Error e -> Lwt_result.fail e
+  in
+  Lwt.on_termination cancelled (fun () ->
+    let rec aux () =
+      if Lwt.is_sleeping proc then (
+         let pp f = Fmt.pf f "jail -r obuilder" in
+         Os.sudo_result ~cwd [ "jail" ; "-r" ; "obuilder" ] ~pp >>= function
+         | Ok () -> Lwt.return_unit
+         | Error (`Msg _) ->
+           Lwt_unix.sleep 10.0 >>= aux
+      ) else Lwt.return_unit (* Process has already finished *)
+    in
+    Lwt.async aux
+  );
+  proc >>= fun r ->
+  copy_log >>= fun () ->
+  if Lwt.is_sleeping cancelled then
+    Lwt.return (r :> (unit, [`Msg of string | `Cancelled]) result)
+  else
+    Lwt_result.fail `Cancelled)
+
+let create ~state_dir:_ _c =
+  Lwt.return {
+    lock = Lwt_mutex.create ();
+    (* Compute a unique (accross obuilder instances) name for the jail.
+       Due to the above mutex, only one jail may be started by a given
+       obuilder process, so appending the obuilder pid is enough to
+       guarantee uniqueness. *)
```

Comment:
This needs to handle multiple jails concurrently from the `ocluster-worker` this will need a different approach to generating the jail name and removing the mutex.

Reply:
There have been changes in the original PR to address this. Can you rebase your PR above mine and give this a new try? (let me know when you want me to squash the commits in my PR)